WordPress/Classes/ViewRelated/Blog/Site Picker/BlogList/BlogListViewModel.swift

```diff
@@ -0,0 +1,237 @@
+            return
+        }
+        let isCurrentlyPinned = blog.pinnedDate != nil
+
+        trackPinned(blog: blog)
+        if isCurrentlyPinned {
+            moveRecentPinnedSiteToRemainingSitesIfNeeded(pinnedBlog: blog)
+        }
+
+        blog.pinnedDate = isCurrentlyPinned ? nil : Date()
+
+        contextManager.saveContextAndWait(contextManager.mainContext)
+    }
+
+    func siteSelected(siteID: NSNumber?) {
+        guard let siteID, let blog = allBlogs.first(where: { $0.dotComID == siteID }) else {
+            return
+        }
+
+        trackSiteSelected(blog: blog)
+
+        blog.lastUsed = Date()
+
+        updateExcessRecentBlogsIfNeeded(selectedSiteID: siteID)
+
+        contextManager.saveContextAndWait(contextManager.mainContext)
+    }
+
+    private func trackPinned(blog: Blog) {
+        eventTracker.track(
+            .siteSwitcherPinUpdated,
+            properties: [
+                "blog_id": blog.dotComID ?? "",
+                "pinned": blog.pinnedDate == nil
+            ]
+        )
+    }
+
+    private func trackSiteSelected(blog: Blog) {
+        let sectionName: String
+
+        if blog.pinnedDate != nil {
+            sectionName = "pinned"
+        } else if blog.lastUsed != nil {
+            sectionName = "recent"
+        } else {
+            sectionName = "all"
+        }
+
+        eventTracker.track(
+            .siteSwitcherSiteTapped,
+            properties: [
+                "section": sectionName,
+            ]
+        )
+    }
+
+    private static func filteredAllRemainingSites(allBlogs: [Blog]) -> [BlogListView.Site] {
+        allBlogs.filter({ $0.pinnedDate == nil && $0.lastUsed == nil }).compactMap(BlogListView.Site.init)
+    }
+
+    private func moveRecentPinnedSiteToRemainingSitesIfNeeded(pinnedBlog: Blog) {
+        if let recentBlogs = recentSitesController?.fetchedObjects,
+           recentBlogs.count == 8 {
+            pinnedBlog.lastUsed = nil
+        }
+    }
+
+    private func updateExcessRecentBlogsIfNeeded(selectedSiteID: NSNumber) {
+        if let recentBlogs = recentSitesController?.fetchedObjects,
+           recentBlogs.count == 8,
+           let lastBlog = recentBlogs.last,
+           !recentBlogs.compactMap({ $0.dotComID }).contains(selectedSiteID) {
+            lastBlog.lastUsed = nil
+        }
+    }
+
+    func viewAppeared() {
+        if recentSites.isEmpty && pinnedSites.isEmpty {
+            selectedBlog()?.lastUsed = Date()
+        }
+
+        contextManager.save(contextManager.mainContext)
+    }
+
+    private func selectedBlog() -> Blog? {
+        return RootViewCoordinator.sharedPresenter.currentOrLastBlog()
+    }
+}
+
+extension BlogListView.Site {
+    init(blog: Blog) {
+        self.init(
+            id: blog.dotComID,
+            title: blog.title ?? "",
+            domain: blog.url ?? "",
+            imageURL: blog.hasIcon ? URL(string: blog.icon!) : nil
```

Comment:
The force unwrapping could fail if under rare circumstances the `hasIcon` is `true`, yet `blog.icon` is `nil`. This case could never happen, to be extra safe, I'd suggest the following minor change:

```suggestion
            imageURL: blog.hasIcon ? URL(string: blog.icon ?? "") : nil
```